l3_terrain_model_generator/src/plugins/base/processor_plugin.cpp

```diff
@@ -34,27 +63,73 @@ bool ProcessorPlugin::postInitialize(const vigir_generic_params::ParameterSet& p
   return true;
 }
 
-void ProcessorPlugin::process(const Timer& timer, UpdatedHandles& updates, const SensorPlugin* sensor)
+void ProcessorPlugin::process(l3::SharedPtr<ProcessingInfo> info, ProcessChain* caller_chain)
 {
   // consider processing rate
-  if (canProcess(timer.current))
+  if (canProcess(info->timer.current))
   {
     l3::UniqueLock lock(mutex_);
-    last_processed_time_ = timer.current;
+    last_processed_time_ = info->timer.current;
   }
   else
     return;
 
   if (enable_timing_)
     profiler_.start();
 
+  // initialize runner
+  boost::unique_lock<boost::mutex> lock(run_mutex_);
+  is_running_ = true;
+  run_info_ = info;
+  lock.unlock();
+
   // call plugin specific implementation
-  processImpl(timer, updates, sensor);
+  // Run the processImpl function in a separate thread
+  if (run_async_)
+    run_cv_.notify_one();
+  // Run the processImpl function in the main thread
+  else
+    runProcessChain();
 
   if (enable_timing_)
     profiler_.stop();
+}
+
+void ProcessorPlugin::processAsync()
+{
+  while (!exit_)
+  {
+    boost::unique_lock<boost::mutex> lock(run_mutex_);
+    run_cv_.wait(lock, [this]{ return is_running_ || exit_; });
+    runProcessChain();
+  }
+}
+
+void ProcessorPlugin::runProcessChain()
+{
+  if (!run_info_)
+  {
+    ROS_ERROR("[%s] No processing info available!", getName().c_str());
+    return;
+  }
+
+  if (exit_)
```

Comment:
here, exit_ could be instead checked at line 92 to avoid confusion about race issues